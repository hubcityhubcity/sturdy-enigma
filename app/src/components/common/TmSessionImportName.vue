<template>
  <SessionFrame>
    <TmFormStruct :submit="onSubmit">
      <h2 class="session-title">
        Choose name
      </h2>
      <div class="session-main bottom-indent">
        <Steps :steps="[`Recover`, `Name`, `Password`]" active-step="Name" />
        <TmFormGroup field-id="import-name" field-label="Your Address">
          <img
            v-if="importedAddress === undefined"
            class="tm-data-msg__icon"
            src="~assets/images/loader.svg"
            alt="a small spinning circle to display loading"
          />
          <p v-else class="address">{{ importedAddress }}</p>
        </TmFormGroup>
        <TmFormGroup
          :error="$v.$error && $v.name.$invalid"
          field-id="import-name"
          field-label="Account Name"
        >
          <TmField
            id="import-name"
            v-model.trim="name"
            type="text"
            placeholder="Must have at least 3 characters"
            vue-focus="vue-focus"
          />
          <TmFormMsg
            v-if="$v.name.$error && !$v.name.required"
            name="Name"
            type="required"
          />
          <TmFormMsg
            v-if="$v.name.$error && !$v.name.minLength"
            name="Name"
            type="minLength"
            min="3"
          />
          <TmFormMsg
            v-if="$v.name.$error && !$v.name.nameExists"
            name="Name"
            type="custom"
            msg="already exists"
          />
        </TmFormGroup>
      </div>
      <div class="session-footer">
        <TmBtn value="Next" type="submit" />
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { required, minLength } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
import SessionFrame from "common/SessionFrame"
import { mapGetters } from "vuex"
import Steps from "../../ActionModal/components/Steps"
import { getWalletIndex } from "@lunie/cosmos-keys"

const nameExists = (value) => {
  const walletIndex = getWalletIndex()
  if (walletIndex.some((e) => e.name === value)) {
    return false
  } else {
    return true
  }
}

export default {
  name: `session-import-name`,
  components: {
    TmBtn,
    TmField,
    SessionFrame,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
    Steps,
  },
  data: () => ({
    importedAddress: undefined,
  }),
  computed: {
    ...mapGetters([`recover`]),
    ...mapGetters([`network`]),
    name: {
      get() {
        return this.$store.state.recover.name
      },
      set(value) {
        this.$store.commit(`updateField`, { field: `name`, value })
      },
    },
  },
  async created() {
    this.importedAddress = await this.$store.dispatch(`getAddressFromSeed`, {
      seedPhrase: this.$store.state.recover.seed,
      network: this.network,
    })
  },
  methods: {
    onSubmit() {
      this.$v.$touch()
      if (this.$v.name.$invalid) return
      this.$router.push("/recover/password")
    },
  },
  validations: () => ({
    name: { required, minLength: minLength(3), nameExists },
  }),
}
</script>
<style scoped>
.address {
  word-break: break-all;
  font-size: 0.9rem;
  color: var(--txt);
}
</style>
