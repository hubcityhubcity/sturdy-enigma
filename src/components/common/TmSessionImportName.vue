<template>
  <SessionFrame>
    <TmFormStruct :submit="onSubmit">
      <h2 class="session-title">
        Choose name
      </h2>
      <div class="session-main bottom-indent">
        <Steps :steps="[`Recover`, `Name`, `Password`]" active-step="Name" />
        <TmFormGroup field-id="import-name" field-label="Your Address">
          <p class="address">{{ importCosmosAddress }}</p>
        </TmFormGroup>
        <TmFormGroup
          :error="$v.$error && $v.name.$invalid"
          field-id="import-name"
          field-label="Account Name"
        >
          <TmField
            id="import-name"
            v-model.trim="name"
            type="text"
            placeholder="Must have at least 3 characters"
            vue-focus="vue-focus"
          />
          <TmFormMsg
            v-if="$v.name.$error && !$v.name.required"
            name="Name"
            type="required"
          />
          <TmFormMsg
            v-if="$v.name.$error && !$v.name.minLength"
            name="Name"
            type="minLength"
            min="3"
          />
          <TmFormMsg
            v-if="$v.name.$error && !$v.name.nameExists"
            name="Name"
            type="custom"
            msg="already exists"
          />
        </TmFormGroup>
      </div>
      <div class="session-footer">
        <TmBtn value="Next" type="submit" />
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { required, minLength } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
import SessionFrame from "common/SessionFrame"
import { mapGetters } from "vuex"
import Steps from "../../ActionModal/components/Steps"
import { getWalletIndex } from "@lunie/cosmos-keys"
import gql from "graphql-tag"

const nameExists = value => {
  const walletIndex = getWalletIndex()
  if (walletIndex.some(e => e.name === value)) {
    return false
  } else {
    return true
  }
}

export default {
  name: `session-import-name`,
  components: {
    TmBtn,
    TmField,
    SessionFrame,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
    Steps
  },
  data: () => ({
    importCosmosAddress: {},
    addressPrefixes: []
  }),
  computed: {
    ...mapGetters([`connected`, `recover`]),
    ...mapGetters({ networkId: `network` }),
    name: {
      get() {
        return this.$store.state.recover.name
      },
      set(value) {
        this.$store.commit(`updateField`, { field: `name`, value })
      }
    }
  },
  async created() {
    // needs to load addresses before using
    await this.$apollo.queries.addressPrefixes.refetch()
    this.importCosmosAddress = await this.$store.dispatch(
      `getAddressFromSeed`,
      {
        seedPhrase: this.$store.state.recover.seed,
        network: this.networkId
      }
    )
  },
  apollo: {
    addressPrefixes: {
      query: gql`
        query Network {
          networks {
            id
            address_prefix
          }
        }
      `,
      /* istanbul ignore next */
      update(data) {
        return data.networks
      },
      prefetch: "true",
      fetchPolicy: "cache-first"
    }
  },
  methods: {
    onSubmit() {
      this.$v.$touch()
      if (this.$v.name.$invalid) return
      this.$router.push("/recover/password")
    }
  },
  validations: () => ({
    name: { required, minLength: minLength(3), nameExists }
  })
}
</script>
<style scoped>
.address {
  word-break: break-all;
  font-size: 0.9rem;
}
</style>
