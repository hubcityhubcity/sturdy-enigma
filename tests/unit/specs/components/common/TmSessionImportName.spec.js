import Vuex from "vuex"
import Vuelidate from "vuelidate"
import { shallowMount, createLocalVue } from "@vue/test-utils"
import TmSessionImportName from "common/TmSessionImportName"
jest.mock(`scripts/google-analytics.js`, () => () => {})
jest.mock("@lunie/cosmos-keys", () => ({
  getWalletIndex: function() {
    return [{ name: `Happy Lunie User`, address: `xyz123` }]
  }
}))
const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuelidate)
localVue.directive(`tooltip`, () => {})
localVue.directive(`focus`, () => {})

describe(`TmSessionImportName`, () => {
  let wrapper, $store, getters

  beforeEach(() => {
    getters = {
      connected: () => true
    }
    $store = {
      state: {
        recover: {
          name: ``,
          seed: ``,
          password: ``,
          passwordConfirm: ``,
          warning: false
        }
      },
      getters,
      commit: jest.fn(),
      dispatch: jest.fn(async () => true),
      mutations: {
        updateField: jest.fn()
      }
    }
    wrapper = shallowMount(TmSessionImportName, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn()
        }
      },
      stubs: [`router-link`]
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`validation should fail if name is not filled in`, async () => {
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.name.$error).toBe(true)
  })

  it(`validation should fail if name lenght < 3 characters`, async () => {
    wrapper.vm.$store.state.recover.name = `as`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.name.$error).toBe(true)
  })

  it(`validation should not fail if name lenght >= 3 characters`, async () => {
    wrapper.vm.$store.state.recover.name = `Happy Lunie User 2`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.name.$error).toBe(false)
  })

  it(`validation should fail if name exists already in stored accounts`, async () => {
    wrapper.vm.$store.state.recover.name = `Happy Lunie User`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.name.$error).toBe(true)
  })

  it(`should commit updateField on field change`, async () => {
    wrapper.setData({ name: `Happy Lunie User` })
    expect($store.commit).toHaveBeenCalledWith(`updateField`, {
      field: `name`,
      value: `Happy Lunie User`
    })
  })

  it(`should go to /recover/password when submit the form`, async () => {
    wrapper.vm.$store.state.recover.name = `Happy Lunie User 2`
    wrapper.vm.onSubmit()
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith(`/recover/password`)
  })
})
