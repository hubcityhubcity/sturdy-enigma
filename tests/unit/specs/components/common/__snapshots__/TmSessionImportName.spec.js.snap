// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionImportName has the expected html structure 1`] = `
<sessionframe-stub>
  <tmformstruct-stub
    submit="function () { [native code] }"
    width=""
  >
    <h2
      class="session-title"
    >
      
      Choose name
    
    </h2>
     
    <div
      class="session-main"
    >
      <steps-stub
        activestep="Name"
        steps="Recover,Name,Password"
      />
       
      <tmformgroup-stub
        fieldid="import-name"
        fieldlabel="Your Address"
      >
        <p
          class="address"
        >
          true
        </p>
      </tmformgroup-stub>
       
      <tmformgroup-stub
        fieldid="import-name"
        fieldlabel="Account Name"
      >
        <tmfield-stub
          id="import-name"
          placeholder="Must have at least 3 characters"
          type="text"
          value=""
          vue-focus="vue-focus"
        />
         
        <!---->
         
        <!---->
         
        <!---->
      </tmformgroup-stub>
    </div>
     
    <div
      class="session-footer"
    >
      <tmbtn-stub
        type="submit"
        value="Next"
      />
    </div>
  </tmformstruct-stub>
</sessionframe-stub>
`;
